predicators/envs/sandwich.py

```
@@ -0,0 +1,832 @@
+            return next_state
+        # Execute pick
+        next_state.set(ing, "pose_x", x)
+        next_state.set(ing, "pose_y", y)
+        next_state.set(ing, "pose_z", self.pick_z)
+        next_state.set(ing, "held", 1.0)
+        next_state.set(self._robot, "fingers", 0.0)  # close fingers
+        return next_state
+
+    def _transition_putonboard(self, state: State, x: float,
+                               y: float) -> State:
+        next_state = state.copy()
+        # Can only putonboard if fingers are closed.
+        if self._GripperOpen_holds(state, [self._robot]):
+            return next_state
+        ing = self._get_held_object(state)
+        assert ing is not None
+        # Check that nothing is on the board.
+        if not self._BoardClear_holds(state, [self._board]):
+            return next_state
+        # Execute putonboard. Rotate object.
+        thickness = self.ingredient_thickness / 2 + self.board_thickness
+        pose_z = self.table_height + thickness
+        next_state.set(ing, "pose_x", x)
+        next_state.set(ing, "pose_y", y)
+        next_state.set(ing, "pose_z", pose_z)
+        next_state.set(ing, "rot", 0.0)
+        next_state.set(ing, "held", 0.0)
+        next_state.set(self._robot, "fingers", 1.0)  # open fingers
+        return next_state
+
+    def _transition_stack(self, state: State, x: float, y: float,
+                          z: float) -> State:
+        next_state = state.copy()
+        # Can only stack if fingers are closed
+        if self._GripperOpen_holds(state, [self._robot]):
+            return next_state
+        # Check that both objects exist
+        ing = self._get_held_object(state)
+        assert ing is not None
+        other_ing = self._get_highest_object_below(state, x, y, z)
+        if other_ing is None:  # no object to stack onto
+            return next_state
+        # Can't stack onto yourself!
+        if ing == other_ing:
+            return next_state
```

do we want to assert this? This should never happen right? Asserting sounds useful to avoid silent failures.

---

it's possible that a random options approach, or some other naive approach, would attempt to use an option that would stack an ingredient onto itself